waiter/docs/service-references.md

```diff
@@ -0,0 +1,78 @@
+# Service References
+
+To process a http request, Waiter needs to resolve a request to a service description.
+This service description tells Waiter how to locate the service instance to use to process the current request.
+
+Waiter uses request headers to construct the service description using the `ServiceDescriptionBuilder`.
+It is possible for the service to be reachable from various combinations of headers:
+  1. token - single or multiple
+  1. on-the-fly services without tokens
+  1. on-the-fly services that extend tokens, e.g. run-as-requester
+
+## Service GC
+
+By default, Waiter relies on the service idle timeout to GC services after periods of inactivity (not receiving requests).
+However, services reachable only via references, e.g. tokens, can be GC-ed eagerly if the reference has been updated.
+
+When a service description is constructed from a request, the service references are also updated.
+These references are available as the `:reference-type->entry` key in the descriptor.
+The `reference-type->entry` is a map where the keys represent a `:type` parameter.
+
+### Reference (`:reference-type->entry` in the descriptor) examples
+
+- A service created with on-the-fly header without tokens will have an empty map
+  as the following value for `:reference-type->entry` in the descriptor:
+```
+  {}
+```
+
+- A service created with `x-waiter-token: foo` header on a request will have the
+  following value for `:reference-type->entry` in the descriptor:
+```
+  {:token {:sources [{:token "foo" :version "v0"}]}}
```

Comment:
If this service is later referenced entirely on the fly, how would this map change?

Reply:
`:reference-type->entry` is computed per request. In the on-the-fly scenario, `:reference-type->entry` would be an empty map `{}`.

Reply:
The references for the service would be a set with two entries: `{}` and `{:token {:sources [{:token "foo" :version "v0"}]}}`.